Add retina_flow_logs_enabled support for azure kubernetes oms_agent  addon profile

wanlonghenry opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description
Add container network logs support for omsagent terraform, this is a new feature and we would want customer to be able to call API with terraform to enable this feature. Feature is required to be registered to test this new property. https://review.learn.microsoft.com/en-us/azure/aks/how-to-configure-container-network-logs?branch=main&branchFallbackFrom=pr-en-us-1274&tabs=cli%2Ccilium
<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 
<img width="1908" height="778" alt="image" src="https://github.com/user-attachments/assets/babd5506-4482-4981-a31b-9c03ec97f8ec" />

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->


## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [x] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->
<img width="1190" height="399" alt="image" src="https://github.com/user-attachments/assets/b599c92b-db63-4a86-b04a-69bdc63f66ec" />

<img width="1861" height="717" alt="image" src="https://github.com/user-attachments/assets/bfa9d8ae-d609-468c-942c-466789f22cda" />

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-merging.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_resource` - support for the `thing1` property [GH-00000]


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)

Fixes #0000


## AI Assistance Disclosure

<!-- 
IMPORTANT!

If you are using any kind of AI/LLM assistance to contribute to the AzureRM provider, this must be disclosed in the pull request.

If this is the case, please check the box below, and include the extent to which AI was used. (e.g. documentation only, code generation, etc.)

If responses to this pull request are/will be generated using AI, disclose this as well.
-->

- [ ] AI Assisted - This contribution was made by, or with the assistance of, AI/LLMs

<!-- extent of AI usage can be described here -->

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

wanlonghenry committed:
add container network logs support for terraform

wanlonghenry committed:
Merge branch 'hashicorp:main' into main

wanlonghenry committed:
update template var name

ms-henglu reviewed (changes requested):
Just one comment about the test, the rest LGTM.

wanlonghenry committed:
add test for retina enable and disable

wanlonghenry committed:
update test

magodo commented on examples/kubernetes/monitoring-log-analytics/main.tf:
```diff
@@ -49,5 +49,6 @@ resource "azurerm_kubernetes_cluster" "example" {
   oms_agent {
     log_analytics_workspace_id      = azurerm_log_analytics_workspace.example.id
     msi_auth_for_monitoring_enabled = true
+    retina_flow_logs_enabled        = false
```
Since this attribute has default value `false`, for the sake of simiplicity of the example, we shall remove this line.

(PS: please leave every comment to the reviewer to resolve, thx!)

magodo commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -275,6 +275,8 @@ An `oms_agent` block exports the following:
 
 * `msi_auth_for_monitoring_enabled` - Is managed identity authentication for monitoring enabled?
 
+* `retina_flow_logs_enabled` - Is Container Network Logs collection enabled?
```
```suggestion
* `retina_flow_logs_enabled` - Is Container Network Logs collection enabled? Defaults to `false`.
```

magodo commented on website/docs/r/kubernetes_cluster.html.markdown:
```diff
@@ -774,6 +774,8 @@ An `oms_agent` block supports the following:
 
 * `msi_auth_for_monitoring_enabled` - (Optional) Is managed identity authentication for monitoring enabled?
 
+* `retina_flow_logs_enabled` - (Optional) Is Container Network Logs collection enabled?
```
Same question for the naming here.

magodo commented on internal/services/containers/kubernetes_addons.go:
```diff
@@ -511,11 +521,16 @@ func flattenKubernetesAddOns(profile map[string]managedclusters.ManagedClusterAd
 			useAADAuth = true
 		}
 
+		if v := kubernetesAddonProfilelocateInConfig(omsAgent.Config, "enableRetinaNetworkFlags"); v != "false" && v != "" {
```
Why not simply test against `true`?

magodo reviewed (changes requested):
Thanks for this PR!
I've taken a look through and left some comments inline. Meanwhile, could you run the acctests locally and attach a screenshot shows the tests are passing?

wanlonghenry committed:
Apply suggestion from @magodo

Co-authored-by: magodo <wztdyl@sina.com>

wanlonghenry commented on internal/services/containers/kubernetes_cluster_data_source.go:
```diff
@@ -1052,11 +1057,16 @@ func flattenKubernetesClusterDataSourceAddOns(profile map[string]managedclusters
 			useAADAuth = true
 		}
 
+		if v := kubernetesAddonProfilelocateInConfig(omsAgent.Config, "enableRetinaNetworkFlags"); v != "false" && v != "" {
```
changed to test againt true

wanlonghenry committed:
address comment

wanlonghenry committed:
Merge branch 'main' of https://github.com/wanlonghenry/terraform-provider-azurerm

wanlonghenry committed:
update descrip

wanlonghenry commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -275,6 +275,8 @@ An `oms_agent` block exports the following:
 
 * `msi_auth_for_monitoring_enabled` - Is managed identity authentication for monitoring enabled?
 
+* `retina_flow_logs_enabled` - Is Container Network Logs collection enabled?
```
updated the description

magodo commented on internal/services/containers/kubernetes_cluster_data_source.go:
```diff
@@ -336,6 +336,10 @@ func dataSourceKubernetesCluster() *pluginsdk.Resource {
 							Type:     pluginsdk.TypeBool,
 							Computed: true,
 						},
+						"retina_flow_logs_enabled": {
```
One more comment is that can we also update the test case for this data source to verifying this attribute?

magodo reviewed (changes requested):
Thanks for this PR!
I have left some additional mostly minor comments that once addressed this should be good to merge 👍

wanlonghenry committed:
address comment for datasource check

wanlonghenry committed:
update test

magodo commented:
Could you mention that a feature is required to be registered to test this new property in the PR description? Given this, could you also pass the acctest and paste a screenshot here as a proof (as in your description the screenshot only convers the unit tests)?

wanlonghenry commented:
> @wanlonghenry Regarding your last commit, what's the motivation?
> 
> [fab0400](https://github.com/hashicorp/terraform-provider-azurerm/commit/fab0400c9910f9fb7ccc8853580388c8c365c000)
> 
> Also, for the description, can you be specific about which feature to register?

Updated the description with document link. 
That is for addressing the scanner check, since I added https://github.com/hashicorp/terraform-provider-azurerm/pull/31256/changes/c389419e07bdbc9835f831a94d61693cff3053e2  the scanner was reporting the issue with  os.Getenv("ARM_CLIENT_ID") which is already deprecated, so I removed this from the test as the test is also skipped and deprecated.

magodo commented:
You shall not just delete them, but follow the CI error message to fix it.

For the acctest, it is the PR author's responsibility to ensure those tests can pass. FWIW, your team can apply for a tenant for test purpose in this case, or you can temporarily reference https://github.com/hashicorp/terraform-provider-azurerm/blob/b0b8bcb74ea2882e55cadc3e8cc14103c6aced96/internal/services/aadb2c/aadb2c_directory_resource_test.go#L120-L134 to use your user account (via Azure CLI) to run the acctest, and then revert the temporary changes back.

wanlonghenry committed:
revert test bac

wanlonghenry committed:
update

wanlonghenry committed:
Merge branch 'hashicorp:main' into main

ms-henglu commented:
<img width="1113" height="440" alt="image" src="https://github.com/user-attachments/assets/6210bb49-1177-4c9a-9002-3791508f50fe" />


I helped run the tests in teamcity, if you want to test it locally, please follow: https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/running-the-tests.md

wanlonghenry committed:
update test

wanlonghenry committed:
Merge branch 'main' of https://github.com/wanlonghenry/terraform-provider-azurerm

wanlonghenry commented:
> Hi @wanlonghenry - This looks like a preview feature based on the linked documentation and the requirement to register this feature within a subscription to be able to use it.
> 
> Could you confirm?

HI @sreallymatt thanks for checking, we are planning to announce GA for this feature and it would also require terraform support. We already supported the cli enablement without requiring register and the doc will be updated together when we announce GA.

sreallymatt reviewed (changes requested):
Hi @wanlonghenry - the added test is failing due to a missing argument, could you update it?

```
    testcase.go:363: Step 1/6 error: Error running pre-apply plan: exit status 1
        Error: Insufficient node_provisioning_profile blocks
          on terraform_plugin_test.tf line 64, in resource "azurerm_kubernetes_cluster" "test":
          64: resource "azurerm_kubernetes_cluster" "test" {
        At least 1 "node_provisioning_profile" blocks are required.
--- FAIL: TestAccKubernetesCluster_addonProfileOMSRetinaFlowLogsToggle (5.05s)
```

sreallymatt commented:
superseded by #33222
